### ISSUE TITLE
Build failure with system installed libraries: alure, clipper, poly2tri or tomcrypt

### ISSUE THREAD

**Issue by akien-mga** [bitfighter/bitfighter#639]
**BF version:** `master` (c63195a175b76d03853976276c922a4d1fa757e1) and `021`. `019f` works fine.
**OS/arch:** Mageia Cauldron, x86_64

**Describe the bug**
I'm trying to update my Mageia package for bitfighter, but I run into a CMake issue whenever either of those system dependencies are installed:
- alure
- clipper
- poly2tri
- tomcrypt (though for the latter I need to use this patch to detect it properly on Mageia: http://svnweb.mageia.org/packages/cauldron/bitfighter/current/SOURCES/bitfighter-019d-mga-system-tomcrypt.patch?revision=808283&view=markup )

Here's the output:
```
$ cd build
$ cmake ..
-- The C compiler identification is GNU 10.2.0
-- The CXX compiler identification is GNU 10.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- CMAKE_DATA_PATH: /usr/local/share.  Change this by invoking cmake with -DCMAKE_DATA_PATH=<SOME_DIRECTORY>
-- Found VorbisFile: optimized;/usr/lib64/libvorbisfile.so;debug;/usr/lib64/libvorbisfile.so  
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found SDL2: /usr/lib64/libSDL2.so;-lpthread  
-- Determining system GL provider
-- Atempting to use standard OpenGL
CMake Warning (dev) at /usr/share/cmake/Modules/FindOpenGL.cmake:305 (message):
  Policy CMP0072 is not set: FindOpenGL prefers GLVND by default when
  available.  Run "cmake --help-policy CMP0072" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  FindOpenGL found both a legacy GL library:

    OPENGL_gl_LIBRARY: /usr/lib64/libGL.so

  and GLVND libraries for OpenGL and GLX:

    OPENGL_opengl_LIBRARY: /usr/lib64/libOpenGL.so
    OPENGL_glx_LIBRARY: /usr/lib64/libGLX.so

  OpenGL_GL_PREFERENCE has not been set to "GLVND" or "LEGACY", so for
  compatibility with CMake 3.10 and below the legacy GL library will be used.
Call Stack (most recent call first):
  CMakeLists.txt:68 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found OpenGL: /usr/lib64/libOpenGL.so   
-- Found ZLIB: /usr/lib64/libz.so (found version "1.2.11") 
-- Found PNG: /usr/lib64/libpng.so (found version "1.6.37") 
-- Found MySQL: /usr/lib64/libmysqlclient.so  
-- Found OGG: optimized;/usr/lib64/libogg.so;debug;/usr/lib64/libogg.so  
-- Found Speex: /usr/lib64/libspeex.so  
-- Found Vorbis: optimized;/usr/lib64/libvorbis.so;debug;/usr/lib64/libvorbis.so  
-- Found ModPlug: /usr/lib64/libmodplug.so  
-- Found OpenAL: /usr/lib64/libopenal.so  
-- Found ALURE: /usr/lib64/libalure.so  
-- Found PkgConfig: /usr/bin/pkg-config (found version "1.7.3") 
-- Checking for module 'sqlite3'
--   Found sqlite3, version 3.33.0
-- Found Sqlite3: /usr/lib64/libsqlite3.so
-- Found TomCrypt: /usr/lib64/libtomcrypt.so  
-- Found Boost: /usr/lib64/cmake/Boost-1.73.0/BoostConfig.cmake (found version "1.73.0")  
-- Found Clipper: /usr/lib64/libpolyclipping.so  
-- Found Poly2tri: /usr/lib64/libpoly2tri.so  
-- Using in-tree LuaJIT
-- Bitfighter version: 021
-- Bitfighter build version: 12030
-- Configuring done
CMake Error at master/CMakeLists.txt:89 (add_dependencies):
  The dependency target "mysql++" of target "master_lib" does not exist.


CMake Error at master/CMakeLists.txt:89 (add_dependencies):
  The dependency target "tomcrypt" of target "master_lib" does not exist.


CMake Error at zap/bitfighter_client.cmake:14 (add_dependencies):
  The dependency target "alure" of target "bitfighter_client" does not exist.
Call Stack (most recent call first):
  zap/CMakeLists.txt:263 (include)


CMake Error at zap/bitfighter_client.cmake:14 (add_dependencies):
  The dependency target "clipper" of target "bitfighter_client" does not
  exist.
Call Stack (most recent call first):
  zap/CMakeLists.txt:263 (include)


CMake Error at zap/bitfighter_client.cmake:14 (add_dependencies):
  The dependency target "poly2tri" of target "bitfighter_client" does not
  exist.
Call Stack (most recent call first):
  zap/CMakeLists.txt:263 (include)


CMake Error at zap/bitfighter_client.cmake:14 (add_dependencies):
  The dependency target "tomcrypt" of target "bitfighter_client" does not
  exist.
Call Stack (most recent call first):
  zap/CMakeLists.txt:263 (include)


CMake Error at zap/bitfighterd.cmake:11 (add_dependencies):
  The dependency target "clipper" of target "bitfighterd" does not exist.
Call Stack (most recent call first):
  zap/CMakeLists.txt:260 (include)


CMake Error at zap/bitfighterd.cmake:11 (add_dependencies):
  The dependency target "poly2tri" of target "bitfighterd" does not exist.
Call Stack (most recent call first):
  zap/CMakeLists.txt:260 (include)


CMake Error at zap/bitfighterd.cmake:11 (add_dependencies):
  The dependency target "tomcrypt" of target "bitfighterd" does not exist.
Call Stack (most recent call first):
  zap/CMakeLists.txt:260 (include)


-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

Removing the above mentioned libraries from the system fixes the CMake errors.
~I also get that issue on my previously packaged version 019f, so I think it's likely related to a CMake change (I have version 3.18.4).~ Actually no, I just retried and 019f builds fine.

**To Reproduce**
- Install distro-provided devel packages for any of the dependencies mentioned above
- Attempt a clean build with cmake, possibly only reproducible with a recent version

**Other**
[CMakeCache.txt](https://github.com/bitfighter/bitfighter/files/5425736/CMakeCache.txt)

**Comment by raptor** [comment]
Hi!

There was some CMake issues at `021` release with the dependencies. Try this patch (from openSUSE):
[cmake3-fixes.zip](https://github.com/bitfighter/bitfighter/files/5425876/cmake3-fixes.zip)

**Comment by akien-mga** [comment]
Thanks. Sadly I still get the same issues after applying that patch.

I amended my post as I had gotten a couple details wrong:

- I tried latest `master` and it has the same issue
- `021` is affected
- `019f` (the previous version I had packaged) works fine
- I didn't try intermediate versions but I somehow remember that I failed to package `019g` back when it was released, possibly for the same reason

**Comment by raptor** [comment]
Try with:

`cmake -DMASTER_MINIMAL=1 ..`

This should reduce errors further. If so, can you repost the cmake output?

**Comment by akien-mga** [comment]
Here's the output with `cmake -DMASTER_MINIMAL=1 ..`:

```
$ cmake -DMASTER_MINIMAL=1 ..
-- The C compiler identification is GNU 10.2.0
-- The CXX compiler identification is GNU 10.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- CMAKE_DATA_PATH: /usr/local/share.  Change this by invoking cmake with -DCMAKE_DATA_PATH=<SOME_DIRECTORY>
-- Found VorbisFile: optimized;/usr/lib64/libvorbisfile.so;debug;/usr/lib64/libvorbisfile.so  
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found SDL2: /usr/lib64/libSDL2.so;-lpthread  
-- Determining system GL provider
-- Atempting to use standard OpenGL
CMake Warning (dev) at /usr/share/cmake/Modules/FindOpenGL.cmake:305 (message):
  Policy CMP0072 is not set: FindOpenGL prefers GLVND by default when
  available.  Run "cmake --help-policy CMP0072" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  FindOpenGL found both a legacy GL library:

    OPENGL_gl_LIBRARY: /usr/lib64/libGL.so

  and GLVND libraries for OpenGL and GLX:

    OPENGL_opengl_LIBRARY: /usr/lib64/libOpenGL.so
    OPENGL_glx_LIBRARY: /usr/lib64/libGLX.so

  OpenGL_GL_PREFERENCE has not been set to "GLVND" or "LEGACY", so for
  compatibility with CMake 3.10 and below the legacy GL library will be used.
Call Stack (most recent call first):
  CMakeLists.txt:68 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found OpenGL: /usr/lib64/libOpenGL.so   
-- Found ZLIB: /usr/lib64/libz.so (found version "1.2.11") 
-- Found PNG: /usr/lib64/libpng.so (found version "1.6.37") 
-- Found MySQL: /usr/lib64/libmysqlclient.so  
-- Found OGG: optimized;/usr/lib64/libogg.so;debug;/usr/lib64/libogg.so  
-- Found Speex: /usr/lib64/libspeex.so  
-- Found Vorbis: optimized;/usr/lib64/libvorbis.so;debug;/usr/lib64/libvorbis.so  
-- Found ModPlug: /usr/lib64/libmodplug.so  
-- Found OpenAL: /usr/lib64/libopenal.so  
-- Found ALURE: /usr/lib64/libalure.so  
-- Found PkgConfig: /usr/bin/pkg-config (found version "1.7.3") 
-- Checking for module 'sqlite3'
--   Found sqlite3, version 3.33.0
-- Found Sqlite3: /usr/lib64/libsqlite3.so
-- Found TomCrypt: /usr/lib64/libtomcrypt.so  
-- Found Boost: /usr/lib64/cmake/Boost-1.73.0/BoostConfig.cmake (found version "1.73.0")  
-- Found Clipper: /usr/lib64/libpolyclipping.so  
-- Found Poly2tri: /usr/lib64/libpoly2tri.so  
-- Using in-tree LuaJIT
-- MASTER_MINIMAL specified.  'master' will only be built with SQLite
-- Bitfighter version: 021
-- Bitfighter build version: 12030
-- Configuring done
CMake Error at zap/bitfighter_client.cmake:25 (add_dependencies):
  The dependency target "clipper" of target "bitfighter_client" does not
  exist.
Call Stack (most recent call first):
  zap/CMakeLists.txt:263 (include)


CMake Error at zap/bitfighter_client.cmake:25 (add_dependencies):
  The dependency target "poly2tri" of target "bitfighter_client" does not
  exist.
Call Stack (most recent call first):
  zap/CMakeLists.txt:263 (include)


CMake Error at zap/bitfighterd.cmake:11 (add_dependencies):
  The dependency target "clipper" of target "bitfighterd" does not exist.
Call Stack (most recent call first):
  zap/CMakeLists.txt:260 (include)


CMake Error at zap/bitfighterd.cmake:11 (add_dependencies):
  The dependency target "poly2tri" of target "bitfighterd" does not exist.
Call Stack (most recent call first):
  zap/CMakeLists.txt:260 (include)


-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

**Comment by raptor** [comment]
Thanks! Was that with, or without that patch I posted earlier?  

I think I know the solution and I will work on another patch.

**Comment by raptor** [comment]
OK, attached is another patch (replaces the old). Can you please test?
[0001-Weird-fix-to-dependencies-when-detected-on-the-syste.zip](https://github.com/bitfighter/bitfighter/files/5427275/0001-Weird-fix-to-dependencies-when-detected-on-the-syste.zip)

**Comment by akien-mga** [comment]
My previous test report was with the `cmake3-fixes.patch` applied.

> OK, attached is another patch (replaces the old). Can you please test?
0001-Weird-fix-to-dependencies-when-detected-on-the-syste.zip

This works great!

**Comment by akien-mga** [comment]
If I set `LUAJIT_BUILTIN=0` to use the system provided luajit, I still get these errors with the last patch:
```
CMake Error at zap/bitfighter_client.cmake:31 (add_dependencies):
  The dependency target "/usr/lib64/libluajit-5.1.so" of target
  "bitfighter_client" does not exist.
Call Stack (most recent call first):
  zap/CMakeLists.txt:263 (include)


CMake Error at zap/bitfighterd.cmake:23 (add_dependencies):
  The dependency target "/usr/lib64/libluajit-5.1.so" of target "bitfighterd"
  does not exist.
Call Stack (most recent call first):
  zap/CMakeLists.txt:260 (include)


```

**Comment by akien-mga** [comment]
With 53a325c3b40616bcbe76990f5b3dfd94e78b87cd and #642 this issue is fixed for me.
I'll let you close it unless you want to keep it around until a patch release is made.

**Comment by raptor** [comment]
Merged #642 

Closing. Thanks!
